lib/src/dependency_data.dart

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:pub_semver/pub_semver.dart';
+
+enum DependencyType { hosted, path, git, sdk }
+
+abstract class DependencyData {
+  DependencyType get type;
+
+  factory DependencyData.fromJson(dynamic data) {
+    if (data == null) {
+      return new HostedData(VersionConstraint.any);
+    } else if (data is String) {
+      return new HostedData(new VersionConstraint.parse(data));
+    } else {
+      var mapData = data as Map;
+
+      var path = mapData['path'] as String;
+      if (path != null) {
+        return new PathData(path);
+      }
+
+      var git = mapData['git'];
+      if (git != null) {
+        return new GitData.fromData(git);
+      }
+
+      final sdk = mapData['sdk'];
+      if (sdk != null) {
+        return new SdkData(sdk);
+      }
+
+      throw new ArgumentError.value(
+          data, 'data', 'No clue how to deal with `$data`.');
```

Comment:
nit: reword the error here for more context, `Unable to understand dependency: $data` or something like that.

Reply:
ack

Reply:
ack